src/Orleans.Core/CodeGeneration/TypeUtils.cs

```diff
@@ -117,23 +118,30 @@ public static string GetTemplatedName(string baseName, Type t, Type[] genericArg
             if (!t.IsGenericType || (t.DeclaringType != null && t.DeclaringType.IsGenericType)) return baseName;
             string s = baseName;
             s += "<";
-            s += GetGenericTypeArgs(genericArguments, fullName);
+            s += GetGenericTypeArgs(t, genericArguments, fullName);
             s += ">";
             return s;
         }
 
-        public static string GetGenericTypeArgs(IEnumerable<Type> args, Predicate<Type> fullName)
+        public static string GetGenericTypeArgs(Type originalType, IEnumerable<Type> args, Predicate<Type> fullName)
         {
             string s = string.Empty;
 
             bool first = true;
+            var genericTypeDefinition = originalType.GetGenericTypeDefinition();
+            var originalGenericArguments = genericTypeDefinition.GetGenericArguments();
             foreach (var genericParameter in args)
             {
                 if (!first)
                 {
                     s += ",";
                 }
-                if (!genericParameter.IsGenericType)
+
+                if (genericParameter.IsGenericParameter)
+                {
+                    s += originalGenericArguments[genericParameter.GenericParameterPosition].Name;
```

Comment:
Should this be calling `GetSimpleTypeName(originalGenericArguments[genericParameter.GenericParameterPosition])` instead?

Reply:
Yes, but because the generic parameter does have any information about the namespace, assembly, full name, etc, I used Name explicitly. Should I change it to call `GetSimpleTypeName`?

Reply:
Done.

Reply:
Thanks, I usually don't bother marking conversations as resolved (maybe for gigantic PRs).